sql/driver/driver_test.go

```diff
@@ -258,3 +259,100 @@ func TestInsecure(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+// concurrentIncrements starts two Goroutines in parallel, both of which
+// read the integer stored at the other's key, increment and update their own.
+// It checks that the outcome is serializable, i.e. exactly one of the
+// two Goroutines (the later write) sees the previous write by the other.
+func concurrentIncrements(db *sql.DB, t *testing.T) {
+	var wgStart, wgEnd sync.WaitGroup
+	wgStart.Add(2)
+	wgEnd.Add(2)
+	for i := 0; i < 2; i++ {
+		go func(i int) {
+			defer wgEnd.Done()
+			shouldStartWait := true
+			// Loop until success.
+			for {
+				txn, err := db.Begin()
+				if err != nil {
+					t.Fatal(err)
+				}
+				// Although the SELECT and the UPDATE below can be combined into a
+				// single statement, we prefer them being separate to ensure that
+				// this transaction is truly a multi-statement transaction, providing
+				// plenty of opportunity to mess up serializability.
+				var value int64
+				err = txn.QueryRow(`SELECT v FROM t.kv WHERE k = $1`, (i+1)%2).Scan(&value)
+				if err != nil {
+					_ = txn.Rollback()
+					continue
+				}
+				// Synchronization point to ensure that there is contention.
+				if shouldStartWait {
```

Comment:
well it's not clear to me whether client test is doing the right thing. I'd like to force the contention. In client test the two go routines wait at the start, but my guess is that once one of them runs, it probably runs until completion (why would it get preempted for the other to run?). It is my intention to rework the other test to be similar to this one, once we've agreed on what to do here. I can include that change in this PR once we agree on what makes sense.

On the use of shouldStartWait, the different error conditions hit continue which retry the transaction and run this logic again. We cannot run this code twice.